src/shogun/machine/gp/Inference.cpp

```diff
@@ -214,105 +207,47 @@ get_negative_log_marginal_likelihood_derivatives(CMap<TParameter*, CSGObject*>*
 	// create lock object
 	CLock lock;
 
-#ifdef HAVE_PTHREAD
-	if (num_deriv<2)
+	#pragma omp parallel for
+	for (index_t i=0; i<num_deriv; i++)
 	{
-#endif /* HAVE_PTHREAD */
-		for (index_t i=0; i<num_deriv; i++)
-		{
-			CMapNode<TParameter*, CSGObject*>* node=params->get_node_ptr(i);
-
-			GRADIENT_THREAD_PARAM thread_params;
+        CMapNode<TParameter*, CSGObject*>* node=params->get_node_ptr(i);
+        SGVector<float64_t> gradient;
+        CLock* locked = &lock;
 
-			thread_params.inf=this;
-			thread_params.obj=node->data;
-			thread_params.param=node->key;
-			thread_params.grad=result;
-			thread_params.lock=&lock;
-
-			get_derivative_helper((void*) &thread_params);
+		if(node->data == this)
+		{
+			// try to find dervative wrt InferenceMethod.parameter
+			gradient=this->get_derivative_wrt_inference_method(node->key);
 		}
-#ifdef HAVE_PTHREAD
-	}
-	else
-	{
-		pthread_t* threads=SG_MALLOC(pthread_t, num_deriv);
-		GRADIENT_THREAD_PARAM* thread_params=SG_MALLOC(GRADIENT_THREAD_PARAM,
-				num_deriv);
-
-		for (index_t t=0; t<num_deriv; t++)
+        else if (node->data == this->m_model)
 		{
-			CMapNode<TParameter*, CSGObject*>* node=params->get_node_ptr(t);
-
-			thread_params[t].inf=this;
-			thread_params[t].obj=node->data;
-			thread_params[t].param=node->key;
-			thread_params[t].grad=result;
-			thread_params[t].lock=&lock;
-
-			pthread_create(&threads[t], NULL, CInference::get_derivative_helper,
-					(void*)&thread_params[t]);
+			// try to find derivative wrt LikelihoodModel.parameter
+			gradient=this->get_derivative_wrt_likelihood_model(node->key);
+		}
+		else if (node->data ==this->m_kernel)
+		{
+			// try to find derivative wrt Kernel.parameter
+			gradient=this->get_derivative_wrt_kernel(node->key);
+		}
+		else if (node->data ==this->m_mean)
+		{
+			// try to find derivative wrt MeanFunction.parameter
+			gradient=this->get_derivative_wrt_mean(node->key);
+		}
+		else
+		{
+			SG_SERROR("Can't compute derivative of negative log marginal "
+					"likelihood wrt %s.%s", node->data->get_name(), node->key->m_name);
 		}
 
-		for (index_t t=0; t<num_deriv; t++)
-			pthread_join(threads[t], NULL);
-
-		SG_FREE(thread_params);
-		SG_FREE(threads);
+		locked->lock();
```

Comment:
#pragma omp critical section would be muc more adequate here, but i would rather refactor this in a lock-free way..... and use maybe `#pragma omp parallel for reduction(...)`